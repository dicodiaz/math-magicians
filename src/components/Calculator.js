import React, { useState } from 'react';
import calculate from '../logic/calculate';
import Button from './Button';

const Calculator = () => {
  const [obj, setObj] = useState({ total: '0', next: null, operation: null });
  const { total, next, operation } = obj;

  const handleClick = (text) => {
    setObj(calculate(obj, text));
  };

  return (
    <div className="col-md-9">
      <div className="row mx-0 row-cols-4 text-center">
        <div className="bg-custom1 text-white text-end col-12 py-4 fs-5">
          {total}
          {operation}
          {next}
        </div>
        <Button classes={['bg-custom2']} text="AC" handleClick={handleClick} />
        <Button classes={['bg-custom2']} text="+/-" handleClick={handleClick} />
        <Button classes={['bg-custom2']} text="%" handleClick={handleClick} />
        <Button classes={['bg-custom3']} text="&divide;" handleClick={handleClick} />
        <Button classes={['bg-custom2']} text="7" handleClick={handleClick} />
        <Button classes={['bg-custom2']} text="8" handleClick={handleClick} />
        <Button classes={['bg-custom2']} text="9" handleClick={handleClick} />
        <Button classes={['bg-custom3']} text="&times;" handleClick={handleClick} />
        <Button classes={['bg-custom2']} text="4" handleClick={handleClick} />
        <Button classes={['bg-custom2']} text="5" handleClick={handleClick} />
        <Button classes={['bg-custom2']} text="6" handleClick={handleClick} />
        <Button classes={['bg-custom3']} text="&minus;" handleClick={handleClick} />
        <Button classes={['bg-custom2']} text="1" handleClick={handleClick} />
        <Button classes={['bg-custom2']} text="2" handleClick={handleClick} />
        <Button classes={['bg-custom2']} text="3" handleClick={handleClick} />
        <Button classes={['bg-custom3']} text="+" handleClick={handleClick} />
        <Button classes={['bg-custom2', 'col-6']} text="0" handleClick={handleClick} />
        <Button classes={['bg-custom2']} text="." handleClick={handleClick} />
        <Button classes={['bg-custom3']} text="=" handleClick={handleClick} />
      </div>
    </div>
  );
};

export default Calculator;
