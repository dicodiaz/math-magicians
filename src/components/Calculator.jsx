import { useState } from 'react';
import buttons from '../data/buttons';
import calculate from '../logic/calculate';
import Button from './Button';

const Calculator = () => {
  const [obj, setObj] = useState({ total: '0', next: null, operation: null });
  const { total, next, operation } = obj;

  const handleClick = (e) => {
    setObj(calculate(obj, e.target.textContent));
  };

  return (
    <div className="col-md-9 pt-5">
      <div className="row mx-0 g-0 row-cols-4 text-center">
        <div data-testid="screen" className="bg-custom1 text-white text-end col-12 p-4 fs-5">
          {total}
          {operation}
          {next}
        </div>
        {buttons.map((button) => {
          const { id, text, type } = button;
          return <Button key={id} text={text} type={type} onClick={handleClick} />;
        })}
      </div>
    </div>
  );
};

export default Calculator;
