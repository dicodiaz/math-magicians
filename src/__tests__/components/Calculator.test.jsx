import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import Calculator from '../../components/Calculator';
import { click } from '../test-utils';

describe('Calculator component', () => {
  it('should render as expected', () => {
    const { container } = render(<Calculator />);
    expect(container).toMatchSnapshot();
  });

  it('should calculate as expected', async () => {
    const tree = render(<Calculator />);
    const screen = tree.getByTestId('screen');
    const equalsButton = tree.getByText('=');
    await click(tree.getByText('2'));
    expect(screen).toHaveTextContent('2');
    await click(tree.getByText('+'));
    await click(tree.getByText('1'));
    await click(equalsButton);
    expect(screen).toHaveTextContent('3');
    await click(tree.getByText('x'));
    await click(tree.getByText('5'));
    await click(equalsButton);
    expect(screen).toHaveTextContent('15');
    await click(tree.getByText('/'));
    await click(tree.getByText('3'));
    await click(equalsButton);
    expect(screen).toHaveTextContent('5');
  });
});
