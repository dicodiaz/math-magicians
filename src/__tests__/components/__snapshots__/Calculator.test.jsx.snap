// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator component should render as expected 1`] = `
<div>
  <div
    class="col-md-9"
  >
    <div
      class="row mx-0 row-cols-4 text-center"
    >
      <div
        class="bg-custom1 text-white text-end col-12 py-4 fs-5"
        data-testid="screen"
      >
        0
      </div>
      <button
        class="py-4 border border-custom1 fs-5 bg-custom2"
        type="button"
      >
        AC
      </button>
      <button
        class="py-4 border border-custom1 fs-5 bg-custom2"
        type="button"
      >
        +/-
      </button>
      <button
        class="py-4 border border-custom1 fs-5 bg-custom2"
        type="button"
      >
        %
      </button>
      <button
        class="py-4 border border-custom1 fs-5 bg-custom3"
        type="button"
      >
        /
      </button>
      <button
        class="py-4 border border-custom1 fs-5 bg-custom2"
        type="button"
      >
        7
      </button>
      <button
        class="py-4 border border-custom1 fs-5 bg-custom2"
        type="button"
      >
        8
      </button>
      <button
        class="py-4 border border-custom1 fs-5 bg-custom2"
        type="button"
      >
        9
      </button>
      <button
        class="py-4 border border-custom1 fs-5 bg-custom3"
        type="button"
      >
        x
      </button>
      <button
        class="py-4 border border-custom1 fs-5 bg-custom2"
        type="button"
      >
        4
      </button>
      <button
        class="py-4 border border-custom1 fs-5 bg-custom2"
        type="button"
      >
        5
      </button>
      <button
        class="py-4 border border-custom1 fs-5 bg-custom2"
        type="button"
      >
        6
      </button>
      <button
        class="py-4 border border-custom1 fs-5 bg-custom3"
        type="button"
      >
        -
      </button>
      <button
        class="py-4 border border-custom1 fs-5 bg-custom2"
        type="button"
      >
        1
      </button>
      <button
        class="py-4 border border-custom1 fs-5 bg-custom2"
        type="button"
      >
        2
      </button>
      <button
        class="py-4 border border-custom1 fs-5 bg-custom2"
        type="button"
      >
        3
      </button>
      <button
        class="py-4 border border-custom1 fs-5 bg-custom3"
        type="button"
      >
        +
      </button>
      <button
        class="py-4 border border-custom1 fs-5 bg-custom2 col-6"
        type="button"
      >
        0
      </button>
      <button
        class="py-4 border border-custom1 fs-5 bg-custom2"
        type="button"
      >
        .
      </button>
      <button
        class="py-4 border border-custom1 fs-5 bg-custom3"
        type="button"
      >
        =
      </button>
    </div>
  </div>
</div>
`;
